operator/pkg/ciliumenvoyconfig/manager.go

```diff
@@ -0,0 +1,231 @@
+	if err != nil {
+		return err
+	}
+
+	if !IsLBProtocolAnnotationEnabled(newObj) && !IsLBProtocolAnnotationEnabled(oldObj) &&
+		!hasAnyPort(newObj, m.ports) && !hasAnyPort(oldObj, m.ports) {
+		return nil
+	}
+	scopedLog.Debug("adding event to queue")
+	m.queue.Add(svcEvent(key))
+	return nil
+}
+
+func (m *manager) OnDeleteService(_ *slim_corev1.Service) error {
+	// Doing nothing here as clean up should be done via k8s OwnerReferences
+	return nil
+}
+
+// run kicks off the controlled loop
+func (m *manager) run(ctx context.Context) {
+	defer m.queue.ShutDown()
+	for {
+		ev, quit := m.queue.Get()
+		if quit {
+			return
+		}
+		err := m.processEvent(ctx, ev)
+		if err != nil {
+			if m.queue.NumRequeues(ev) < m.maxRetries {
+				log.WithError(err).Warning("Error while processing event, retrying")
+				m.queue.AddRateLimited(ev)
+			} else {
+				log.WithError(err).Warning("Error while processing event, no more retries")
+				m.queue.Forget(ev)
+			}
+			continue
+		}
+		m.queue.Done(ev)
+	}
+}
+
+func (m *manager) processEvent(ctx context.Context, event interface{}) error {
+	switch k := event.(type) {
+	case svcEvent:
+		n := string(k) // service namespace/name
+
+		objFromCache, exists, err := m.serviceStore.GetByKey(n)
+		if err != nil {
+			return err
+		}
+		if !exists {
+			return fmt.Errorf("service %s does not exist", n)
+		}
+
+		service, ok := objFromCache.(*slim_corev1.Service)
+		if !ok {
+			return fmt.Errorf("got invalid object from cache: %T", objFromCache)
+		}
+		if IsLBProtocolAnnotationEnabled(service) || hasAnyPort(service, m.ports) {
+			return m.createEnvoyConfig(ctx, service)
+		}
+		return m.deleteEnvoyConfig(ctx, service)
+	default:
+		log.Debugf("Encountered an unknown key type %T in CEC controller", k)
+		return fmt.Errorf("unknown key type %T", k)
+	}
+}
+
+func (m *manager) MarkSynced() {
+	m.envoyConfigManager.MarkSynced()
+}
+
+func (m *manager) createEnvoyConfig(ctx context.Context, svc *slim_corev1.Service) error {
+	desired, err := m.getEnvoyConfigForService(svc)
+	if err != nil {
+		return err
+	}
+
+	// check if the CiliumEnvoyConfig resource already exists
+	key, err := cache.MetaNamespaceKeyFunc(desired)
+	if err != nil {
+		log.WithError(err).Warn("MetaNamespaceKeyFunc returned an error")
+		return err
+	}
+	existingEnvoyConfig, exists, err := m.envoyConfigManager.getByKey(key)
+	if err != nil {
+		log.WithError(err).Warn("CiliumEnvoyConfig lookup failed")
+		return err
+	}
+
+	scopedLog := log.WithField(logfields.ServiceKey, getName(svc))
+	if exists {
+		if desired.DeepEqual(existingEnvoyConfig) {
+			log.WithField(logfields.CiliumEnvoyConfigName, key).Debug("No change for existing CiliumEnvoyConfig")
+			return nil
+		}
+		// Update existing CEC
+		newEnvoyConfig := existingEnvoyConfig.DeepCopy()
+		newEnvoyConfig.Spec = desired.Spec
+		_, err = k8s.CiliumClient().CiliumV2().CiliumEnvoyConfigs(svc.Namespace).Update(context.Background(), newEnvoyConfig, metav1.UpdateOptions{})
```

Comment:
Is it worth doing a Patch here, or is that an unnecessary optimization?

Reply:
I wondered about this, if something else is updating the CEC then this might break that.